### PR TITLE
FSE: Handle Gutenboarding launch button

### PR DESCRIPTION
**Pull request by sirreal**
#### Changes proposed in this Pull Request

Part of https://github.com/Automattic/wp-calypso/pull/42847.
Companion to https://github.com/Automattic/wp-calypso/pull/42849.

Details in https://github.com/Automattic/wp-calypso/pull/42847 and D44235-code.

#### Testing instructions

Sandbox `widgets.wp.com` and `example.wordpress.com` (Simple site).

* Test D44235-code in isolation. It should result in no changes in the block editor.
* Test in combination with D44258-code. It should result in no changes in the block editor.

1. https://wordpress.com/block-editor should work as before.
1. Starting from [`/new`](https://wordpress.com/new), follow the flow through site creation to the editor.
1. In the editor, the publish button should be replaced with "Launch."
1. The launch button should link to a URL like https://wordpress.com/start/new-launch?siteSlug=example.wordpress.com&source=editor

All of the testing is to ensure there are no regressions.

### CODE REVIEW

File: apps/full-site-editing/full-site-editing-plugin/editor-gutenboarding-launch/index.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+	// Avoid loading assets if possible.
+	if ( ! \A8C\FSE\Common\is_block_editor_screen() ) {
+		return;
+	}
+
+	$asset_file          = include plugin_dir_path( __FILE__ ) . 'dist/editor-gutenboarding-launch.asset.php';
+	$script_dependencies = isset( $asset_file['dependences'] ) ? $asset_file['dependencies'] : array();
+	$script_version      = isset( $asset_file['verson'] ) ? $asset_file['version'] : filemtime( plugin_dir_path( __FILE__ ) . 'dist/editor-gutenboarding-launch.js' );
+	$style_version       = isset( $asset_file['verson'] ) ? $asset_file['version'] : filemtime( plugin_dir_path( __FILE__ ) . 'dist/editor-gutenboarding-launch.css' );
```
Comment (alshakero) [inline apps/full-site-editing/full-site-editing-plugin/editor-gutenboarding-launch/index.php:32]: I assume you mean `version`?

File: apps/full-site-editing/full-site-editing-plugin/editor-gutenboarding-launch/index.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+		isFseGutenboarding?: boolean;
+		[ key: string ]: unknown;
+	};
+}
+declare const window: CalypsoifyWindow;
+
+domReady( () => {
+	updateEditor();
+	// Hook fallback incase setGutenboardingStatus runs after initial dom render.
+	addAction( 'setGutenboardingStatus', 'a8c-gutenboarding', updateEditor );
```
Comment (alshakero) [inline apps/full-site-editing/full-site-editing-plugin/editor-gutenboarding-launch/index.ts:25]: Maybe move the condition from `updateEditor`? 

```suggestion
if (
  !window?.calypsoifyGutenberg?.isFseGutenboarding ||
  !window?.calypsoifyGutenberg?.frankenflowUrl
) {
  addAction("setGutenboardingStatus", "a8c-gutenboarding", updateEditor);
} else {
  updateEditor();
}

```
I'm assuming this condition means the hook isn't needed.

This way you won't add the action unless needed.

**Review by alshakero — COMMENTED** [review]
Reviewed the code. Testing now.

File: apps/full-site-editing/full-site-editing-plugin/editor-gutenboarding-launch/index.php (hunk 1 above)
Comment (sirreal) [reply apps/full-site-editing/full-site-editing-plugin/editor-gutenboarding-launch/index.php:32]: Good… we've confirmed the fallback is working as designed 😅

Comment (sirreal) [inline apps/full-site-editing/full-site-editing-plugin/editor-gutenboarding-launch/index.php:32]:
```suggestion
	$script_version      = isset( $asset_file['version'] ) ? $asset_file['version'] : filemtime( plugin_dir_path( __FILE__ ) . 'dist/editor-gutenboarding-launch.js' );
	$style_version       = isset( $asset_file['version'] ) ? $asset_file['version'] : filemtime( plugin_dir_path( __FILE__ ) . 'dist/editor-gutenboarding-launch.css' );
```

File: apps/full-site-editing/full-site-editing-plugin/editor-gutenboarding-launch/index.ts (hunk 2 above)
Comment (sirreal) [reply apps/full-site-editing/full-site-editing-plugin/editor-gutenboarding-launch/index.ts:25]: The values are potentially set between domReady and when the action is fired. The goal is to:

- Run as early as possible
- Run 0 or 1 times

We can move the conditional checks around, but it's important to check them both places.

Comment (alshakero) [reply apps/full-site-editing/full-site-editing-plugin/editor-gutenboarding-launch/index.ts:25]: Then I think as is is better.

**Review by alshakero — APPROVED** [review]
Tested and all works as prescribed.

**Comment by sirreal** [comment]
I cancelled e2e tests which are irrelevant for FSE changes.

File: apps/full-site-editing/full-site-editing-plugin/editor-gutenboarding-launch/index.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+		$script_dependencies,
+		$script_version,
+		true
+	);
+
+	wp_enqueue_style(
+		'a8c-fse-editor-gutenboarding-launch-style',
+		plugins_url( 'dist/editor-gutenboarding-launch.css', __FILE__ ),
+		array(),
+		$style_version,
```
Comment (obenland) [inline apps/full-site-editing/full-site-editing-plugin/editor-gutenboarding-launch/index.php:46]: This trailing comma is causing a fatal error for me locally:
`Parse error: syntax error, unexpected ')' in wp-calypso/apps/full-site-editing/full-site-editing-plugin/editor-gutenboarding-launch/index.php on line 47`

Comment (simison) [reply apps/full-site-editing/full-site-editing-plugin/editor-gutenboarding-launch/index.php:46]: What's your PHP version locally?

This should work with PHP 7.3 I think.

Comment (sirreal) [reply apps/full-site-editing/full-site-editing-plugin/editor-gutenboarding-launch/index.php:46]: Thanks for spotting that, we need to support PHP 7.2:

Remove the trailing comma: https://github.com/Automattic/wp-calypso/pull/42918

**Pull request merged**